include/flux/op/cartesian_base.hpp

```diff
@@ -0,0 +1,361 @@
+            auto our_sz = flux::size(get_base<I>(self));
+            auto our_dist = flux::distance(get_base<I>(self), std::get<I>(from), std::get<I>(to));
+            return prev_dist * our_sz + our_dist;
+        }
+    }
+
+    template <std::size_t I, typename Self, typename Fn>
+    static constexpr auto read1_(Fn fn, Self& self, cursor_t<Self> const& cur)
+    -> decltype(auto)
+    {
+        return fn(get_base<I>(self), std::get<I>(cur));
+    }
+
+    template <typename Fn, typename Self>
+    static constexpr auto read_(Fn& fn, Self& self, cursor_t<Self> const& cur)
+    -> decltype(auto)
+        requires (ReadKind == read_kind::tuple)
+    {
+        return [&]<std::size_t... N>(std::index_sequence<N...>) {
+            return std::tuple<decltype(read1_<N>(fn, self, cur))...>(read1_<N>(fn, self, cur)...);
+        }(std::make_index_sequence<Arity>{});
+    }
+
+    template <std::size_t I, typename Self, typename Function,
+            typename... PartialElements>
+    static constexpr void for_each_while_impl(Self& self,
+                                              bool& keep_going,
+                                              cursor_t<Self>& cur,
+                                              Function&& func,
+                                              PartialElements&&... partial_elements)
+    {
+        // We need to iterate right to left.
+        if constexpr (I == Arity - 1) {
+            std::get<I>(cur) = flux::for_each_while(get_base<I>(self),
+                [&](auto&& elem) {
+                    keep_going = std::invoke(func,
+                                             element_t<Self>(FLUX_FWD(partial_elements)..., FLUX_FWD(elem)));
+                    return keep_going;
+                });
+        } else {
+            std::get<I>(cur) = flux::for_each_while(get_base<I>(self),
+                [&](auto&& elem) {
+                    for_each_while_impl<I+1>(
+                            self, keep_going, cur,
+                            func, FLUX_FWD(partial_elements)..., FLUX_FWD(elem));
+                    return keep_going;
+                });
+        }
+    }
+
+protected:
+    using types = cartesian_traits_types<Arity, CartesianKind, ReadKind, Bases...>;
+
+public:
+
+    template <typename Self>
+    static constexpr auto first(Self& self)
+        requires (CartesianKind == cartesian_kind::product)
+    {
+        return std::apply([](auto&&... args) {
+            return std::tuple(flux::first(FLUX_FWD(args))...);
+        }, self.bases_);
+    }
+
+    template <typename Self>
+    static constexpr auto first(Self& self)
+        requires (CartesianKind == cartesian_kind::power)
+    {
+        return []<std::size_t... Is>(auto&& arg, std::index_sequence<Is...>) {
+            std::array<cursor_t<require_single_type_t<Bases...>>, Arity> cur = {(static_cast<void>(Is), flux::first(arg))...};
+            return cur;
+        }(self.base_, std::make_index_sequence<Arity>{});
```

Comment:
This is going to result in `Arity` calls to `flux::first(self.base_)`, which could result in a long chain of function calls if we're at the end of a long pipeline. But since we know `Base` is multipass, we know that its cursor type is copyable, so we can instead just call `flux::first(self.base_)` once and copy it into the array N times.

(Also, I don't think we need the `require_single_type_t` helper, because we know the `Bases...` pack only has one element in this function?)

```suggestion
        auto base_cur = flux::first(self.base_);
        return [&]<std::size_t... Is>(std::index_sequence<Is...>) {
            std::array<cursor_t<Bases>..., Arity> cur{(static_cast<void>(Is), base_cur)...};
            return cur;
        }(std::make_index_sequence<Arity>{});
```

Reply:
Ah yep, I made the `require_single_type_t` helper specifically to explicitly convey to a reader that we're asserting that the `Bases...` pack only has one element in the function - I thought it better to show intent there and protect against if it was ever accidentally changed. What are your thoughts?

Reply:
I like the intent, but I think this is probably better conveyed by saying `static_assert(sizeof...(Bases) == 1)`, which avoids the extra template instantiations that would be caused by a `require_single_type` class template (and actually lead to a better error message if this does gets screwed up at some point in the future).

Reply:
I see, I hadn't considered the cost of extra template instantiations. I've changed it to the static assert 👍